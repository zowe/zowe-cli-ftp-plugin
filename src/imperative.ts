/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { IImperativeConfig } from "@brightside/imperative";
import { FTPConfig } from "./api/FTPConfig";

const config: IImperativeConfig = {
    commandModuleGlobs: ["**/cli/*/*.definition!(.d).*s"],
    rootCommandDescription: "Data set and job functionality via FTP. This functionality " +
    "uses the open source zos-node-accessor package from IBM. Commands under this group " +
    "require you to create a zftp profile before using them.\n" +
    "If you find this functionality useful, please consider setting up z/OSMF on " +
    "your system to get improved stability and speed and more features (for example, issuing TSO and console commands) by using " +
    "core Zowe CLI.",
    productDisplayName: "z/OS FTP Plugin",
    pluginAliases: ["zftp"],
    pluginSummary: "z/OS Files and jobs via FTP",
    pluginHealthCheck: "./lib/healthCheck.handler",
    name: "zos-ftp",
    profiles: [
        {
            type: "zftp",
            createProfileExamples: [{
                options: "myprofile -u ibmuser -p ibmp4ss -H sys123",
                description: "Create a zftp profile called 'myprofile' with default settings (port" +
                ", timeout, etc.) to connect with the host system 123."
            }],
            schema: {
                type: "object",
                title: "Configuration profile for z/OS FTP",
                description: "Configuration profile for z/OS FTP",
                properties: {
                    host: {
                        type: "string",
                        optionDefinition: FTPConfig.OPTION_HOST
                    },
                    port: {
                        type: "number",
                        optionDefinition: FTPConfig.OPTION_PORT
                    },
                    user: {
                        type: "string",
                        optionDefinition: FTPConfig.OPTION_USER,
                        secure: true
                    },
                    password: {
                        type: "string",
                        optionDefinition: FTPConfig.OPTION_PASSWORD,
                        secure: true
                    },
                    secureFtp: {
                        type: "string",
                        optionDefinition: FTPConfig.OPTION_SECURE_FTP
                    },
                    rejectUnauthorized: {
                        type: ["boolean", "null"],
                        optionDefinition: FTPConfig.OPTION_REJECT_UNAUTHORIZED
                    },
                    servername: {
                        type: ["string", "null"],
                        optionDefinition: FTPConfig.OPTION_SERVER_NAME
                    },
                    connectionTimeout: {
                        type: "number",
                        optionDefinition: FTPConfig.OPTION_CONNECTION_TIMEOUT
                    },
                },
                optional: ["host", "port", "user", "password"],
            },
        },
    ]
};

export = config;
