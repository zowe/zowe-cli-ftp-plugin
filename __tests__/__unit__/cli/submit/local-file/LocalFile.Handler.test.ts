/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { IO } from "@zowe/imperative";
import SubmitJobFromLocalFileHandler from "../../../../../src/cli/submit/local-file/LocalFile.Handler";
import TestUtils from "../../TestUtils";

describe("Submit job from local file handler", () => {

    it("should return correct message if the local file is submit successfully.", async () => {
        const handler = new SubmitJobFromLocalFileHandler();
        const jobDetails = {
            jobname: "jobname1",
            jobid: "jobid1",
            spoolFiles: [
                {
                    name: "file1"
                },
                {
                    name: "file2"
                }
            ]
        };

        IO.readFileSync = jest.fn().mockReturnValue(Promise.resolve("sss"));
        const mockResponse = TestUtils.getMockResponse();
        const mockParams: any = {
            arguments: {
                file: "/u/user/jcl"
            },
            connection: {
                submitJCL: jest.fn().mockReturnValue(Promise.resolve("jobid1")),
                getJobStatus: jest.fn().mockReturnValue(Promise.resolve(jobDetails))
            },
            response: mockResponse
        };
        await handler.processFTP(mockParams);
        expect(mockResponse.data.setObj.mock.calls[0]).toMatchSnapshot();
        expect(mockResponse.format.output.mock.calls[0]).toMatchSnapshot();
    });

    it("should return correct message if the local file is submit successfully with wait option.", async () => {

        const handler = new SubmitJobFromLocalFileHandler();

        const jobDetails = {
            jobid: "jobid2",
            jobname: "jobname2",
            owner: "owner2",
            status: "OUTPUT",
            rc: 0,
            retcode: "CC 0000",
            spoolFiles: [
                {
                    name: "file1"
                },
                {
                    name: "file2"
                }
            ]
        };


        const jobRuning = {
            jobid: "jobid2",
            jobname: "jobname2",
            owner: "owner2",
            status: "ACTIVE"
        };

        IO.readFileSync = jest.fn().mockReturnValue(Promise.resolve("sss"));
        const mockResponse = TestUtils.getMockResponse();
        const mockParams: any = {
            arguments: {
                file: "/u/user/jcl",
                wait: "5,10"
            },
            connection: {
                submitJCL: jest.fn().mockReturnValue(Promise.resolve("jobid2")),
                getJobStatus: jest.fn().mockReturnValueOnce(Promise.resolve(jobRuning)).mockReturnValueOnce(Promise.resolve(jobRuning))
                    .mockReturnValue(Promise.resolve(jobDetails))
            },
            response: mockResponse
        };
        await handler.processFTP(mockParams);
        expect(mockResponse.data.setObj.mock.calls[0][0]).toMatchSnapshot();
        expect(mockResponse.format.output.mock.calls[0][0]).toMatchSnapshot();
        expect(mockParams.connection.getJobStatus).toHaveBeenCalledTimes(3);
        expect(mockResponse.console.log.mock.calls[1][0]).toBe("Waiting for job completion.");
    });

});
