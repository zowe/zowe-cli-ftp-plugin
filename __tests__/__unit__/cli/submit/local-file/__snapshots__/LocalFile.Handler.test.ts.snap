// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Submit job from local file handler should return correct message if the local file is submit successfully with wait option. 1`] = `
Object {
  "jobid": "jobid2",
  "jobname": "jobname2",
  "owner": "owner2",
  "rc": 0,
  "retcode": "CC 0000",
  "spoolFiles": Array [
    Object {
      "name": "file1",
    },
    Object {
      "name": "file2",
    },
  ],
  "status": "OUTPUT",
}
`;

exports[`Submit job from local file handler should return correct message if the local file is submit successfully with wait option. 2`] = `
Object {
  "fields": Array [
    "jobid",
    "jobname",
    "owner",
    "status",
    "rc",
    "retcode",
  ],
  "format": "object",
  "output": Object {
    "jobid": "jobid2",
    "jobname": "jobname2",
    "owner": "owner2",
    "rc": 0,
    "retcode": "CC 0000",
    "spoolFiles": Array [
      Object {
        "name": "file1",
      },
      Object {
        "name": "file2",
      },
    ],
    "status": "OUTPUT",
  },
}
`;

exports[`Submit job from local file handler should return correct message if the local file is submit successfully. 1`] = `
Array [
  Object {
    "jobid": "jobid1",
    "jobname": "jobname1",
    "spoolFiles": Array [
      Object {
        "name": "file1",
      },
      Object {
        "name": "file2",
      },
    ],
  },
]
`;

exports[`Submit job from local file handler should return correct message if the local file is submit successfully. 2`] = `
Array [
  Object {
    "fields": Array [
      "jobid",
      "jobname",
      "owner",
      "status",
    ],
    "format": "object",
    "output": Object {
      "jobid": "jobid1",
      "jobname": "jobname1",
      "spoolFiles": Array [
        Object {
          "name": "file1",
        },
        Object {
          "name": "file2",
        },
      ],
    },
  },
]
`;
